crates/cli/src/lib.rs

```diff
@@ -0,0 +1,339 @@
+                tx.hash, read_tx
+            )
+        })?;
+
+    if let Some(server_jh) = server_jh {
+        let _ = server_jh.await;
+    }
+
+    Ok((tx_hash.to_string(), prover_hash, verifier_hash))
+}
+
+fn create_tx_data(
+    verified_file: Option<PathBuf>,
+    tx_input_args: String,
+    file_url_args: Option<String>,
+    served_file_map: &HashMap<&PathBuf, String>,
+) -> std::result::Result<ProgramMetadata, String> {
+    let (image_file_name, image_file_url, image_file_checksum) =
+        generate_tx_data(verified_file, tx_input_args, file_url_args, served_file_map)?;
+    let mut data = ProgramMetadata {
+        //TODO what is the role of the name?
+        name: image_file_name.clone(),
+        hash: Hash::default(),
+        image_file_name,
+        image_file_url: image_file_url.to_string(),
+        image_file_checksum,
+    };
+    //TODO use a new to be sure the update is done during creation.
+    data.update_hash();
+    Ok(data)
+}
+
+fn generate_tx_data(
+    verified_file: Option<PathBuf>,
+    tx_input_args: String,
+    file_url_args: Option<String>,
+    served_file_map: &HashMap<&PathBuf, String>,
+) -> std::result::Result<(String, String, String), String> {
+    verified_file
+        .and_then(|path| {
+            //case 1 prover is a file name
+            extract_hash_from_file_content(&path)
+                .and_then(|file_hash| {
+                    path.file_name()
+                        .and_then(|file_name| file_name.to_str())
+                        .map(|file_name| (file_name.to_string(), file_hash))
+                })
+                .and_then(|(file_name, file_hash)| {
+                    served_file_map
+                        .get(&path)
+                        .map(|url| Ok((file_name, url.to_string(), file_hash)))
+                })
+        })
+        .or_else(|| {
+            //case 2 prover is a hash
+            //try to detect if the parameters is a hash
+            let res = hex::decode(&tx_input_args)
+                .map_err(|err| format!("error during Hash decoding:{err}"))
+                .and_then(|val| {
+                    <[u8; 32]>::try_from(val)
+                        .map_err(|err| format!("error Hash binary vec into array conv:{err:?}"))?;
+                    Ok(())
+                })
+                .and_then(|_| {
+                    //Get file name from url or define as file hash
+                    let filename = file_url_args
+                        .as_ref()
+                        .and_then(|url| get_last_segment_from_url(url))
+                        .unwrap_or(tx_input_args.to_string());
+                    file_url_args
+                        .map(|img_url| (filename, img_url, tx_input_args))
+                        .ok_or("Image url not provided with Hash".to_string())
+                });
+            Some(res)
+        })
+        //never None
+        .unwrap()
```

Comment:
This is a nice block! Reads very well! :heart_eyes:

Reply:
Thanks, when you have result and option, it's often easier to use the functional style to manage all case in a clear way.